hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/MemberSchemaService.java

```diff
@@ -0,0 +1,188 @@
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.hazelcast.internal.util.InvocationUtil.invokeOnStableClusterSerial;
+
+public class MemberSchemaService implements ManagedService, PreJoinAwareService, SchemaService {
+
+    public static final String SERVICE_NAME = "schema-service";
+    private static final int MAX_RETRIES = 100;
+    private final Map<Long, Schema> schemas = new ConcurrentHashMap<>();
+    private final Set<Schema> seenSchemas = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private ILogger logger;
+    private NodeEngine nodeEngine;
+
+    @Override
+    public void init(NodeEngine nodeEngine, Properties properties) {
+        this.logger = nodeEngine.getLogger(SchemaService.class);
+        this.nodeEngine = nodeEngine;
+    }
+
+    @Override
+    public void reset() {
+    }
+
+    @Override
+    public void shutdown(boolean terminate) {
+        schemas.clear();
+    }
+
+    @Override
+    public Operation getPreJoinOperation() {
+        if (schemas.size() == 0) {
+            return null;
+        }
+        if (logger.isFinestEnabled()) {
+            logger.finest("Preparing prejoin operation with schemas " + schemas);
+        }
+        return new SendAllSchemasOperation(new ArrayList<>(schemas.values()));
+    }
+
+    @Override
+    public Schema get(long schemaId) {
+        return getAsync(schemaId).join();
+    }
+
+    public CompletableFuture<Schema> getAsync(long schemaId) {
+        Schema schema = getLocal(schemaId);
+        if (schema != null) {
+            return CompletableFuture.completedFuture(schema);
+        }
+        if (logger.isFinestEnabled()) {
+            logger.finest("Could not find schema id  " + schemaId + " locally, will search on the cluster" + schemaId);
+        }
+        ClusterService cluster = nodeEngine.getClusterService();
+        OperationService operationService = nodeEngine.getOperationService();
+        Set<Member> members = cluster.getMembers();
+        Iterator<Member> iterator = members.iterator();
+        return searchClusterAsync(schemaId, iterator, operationService);
+    }
+
+    private CompletableFuture<Schema> searchClusterAsync(long schemaId, Iterator<Member> iterator,
+                                                         OperationService operationService) {
+        if (!iterator.hasNext()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        Address address = iterator.next().getAddress();
+        FetchSchemaOperation op = new FetchSchemaOperation(schemaId);
+        InvocationFuture<Schema> future = operationService.invokeOnTarget(SERVICE_NAME, op, address);
```

Comment:
Since this is read, it is enough to read from one node in the common case. 
invokeOnStableCluster is to make sure to that we replicate to all cluster.